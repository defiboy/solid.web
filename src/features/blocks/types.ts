import { Transaction } from "features/transactions";

export interface BlockHeader {
    number: number
    hash: string
    parentHash: string
    nonce: string
    sha3Uncles: string
    logsBloom: string
    transactionRoot: string
    stateRoot: string
    receiptRoot: string
    miner: string
    extraData: string
    gasLimit: number
    gasUsed: number
    timestamp: number | string
}

export interface Block extends BlockHeader {
    transactions: Transaction[];
    size: number
    difficulty: number
    totalDifficulty: number
    uncles: string[];
}