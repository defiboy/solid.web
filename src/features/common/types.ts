export enum Status { // TODO, with sagas I think it won't apply
    InProgress = 'InProgress',
    Completed = 'Completed',
    Synchronizing = 'Synchronizing', // this won't apply for connection modal
    Failed = 'Failed',
    NotStarted = 'NotStarted',
    Started = 'Started'
}

export interface GenericResponse<T> {
    total: number;
    limit: number;
    skip: number;
    data: T;
}

export interface GenericArrayResponse<T> {
    total: number;
    limit: number;
    skip: number;
    data: T[];
}
