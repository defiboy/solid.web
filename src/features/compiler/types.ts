import { Status } from '../common/types'

export interface ValidateSourceCode {
  status: Status
  compilerVersion: string
  sourceCode: string
}

export interface LoadCompilerRequest {
  version: string
  status: Status
}
