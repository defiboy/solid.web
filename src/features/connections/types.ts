import { Connection } from '@solid-explorer/types'

export interface ConnectionItem extends Connection {
  type: string
}
