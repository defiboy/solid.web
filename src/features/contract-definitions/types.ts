import { ContractDefinition } from '@solidstudio/solid.types'

export interface ContractDefinitionItem extends ContractDefinition {
    type: string
}