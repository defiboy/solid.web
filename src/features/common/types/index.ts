export enum Status { // TODO, with sagas I think it won't apply
    InProgress = 'InProgress',
    Completed = 'Completed',
    Synchronizing = 'Synchronizing', // this won't apply for connection modal
    Failed = 'Failed',
    NotStarted = 'NotStarted',
    Started = 'Started'
}