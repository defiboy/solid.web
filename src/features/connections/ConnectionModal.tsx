import React from 'react'
import { bindActionCreators, Dispatch, ActionCreator, Action } from 'redux'
import { connect } from 'react-redux'

import { Form } from 'antd'
import { FormikErrors, Field } from 'formik'

import { AppState } from './reducer'
import { TextField } from '../../components'

import { Status } from "../common/types" // TODO: this shouldn't be the case with Sagas

import { createOrUpdateConnection, createConnectionCancelled } from './actions'
import { Connection, CreateConnection } from './types'
import { ConnectionModalComponent } from "./ConnectionModalComponent";

const FORM_ID = 'CONNECTION_FORM'

interface OwnProps {

}

interface StateProps {
  visible: boolean
  loading: boolean
  submitted: boolean
  createConnection: CreateConnection
}

interface DispatchProps {
  // itemToEdit?: Connection
  createOrUpdateConnection: (item: Connection) => void
  createConnectionCancelled: ActionCreator<Action>
}

type AllProps = OwnProps & DispatchProps & StateProps

export class ConnectionModal extends React.Component<AllProps> {
  // static defaultProps = {
  //   visible: false,
  //   loading: false    
  // }
  saveConnection = (item: Connection) => {
    // if (this.state.itemToEdit) { // TODO
    //     // update values
    // }
    console.log('save connection', item)
    this.props.createOrUpdateConnection(item)
  }

  render() {
    return (
      <ConnectionModalComponent
        formId={FORM_ID}
        title="Add Connection"
        onSubmit={this.saveConnection}
        visible={this.props.visible}
        loading={this.props.loading}
        onCancel={this.props.createConnectionCancelled}
        initialValues={{ name: '13', url: '', transactionReceipts: [] }}
        validator={(items: Connection) => {
          const errors: FormikErrors<Connection> = {}
          if (!items.name) {
            errors.name = 'Required'
          }
          if (!items.url) {
            errors.url = 'Required'
          }
          return errors
        }}
        FormComponent={({ fields: { name, url }, onSubmit }) => (
          <Form id={FORM_ID} onSubmit={onSubmit}>
            <Field
              name="name"
              render={(innerProps: any) => <TextField {...innerProps} label="Name" placeHolder="http://" />}
            />
            <Field
              name="url"
              render={(innerProps: any) => (
                <TextField {...innerProps} label="Blockchain URL" placeHolder="JSON RPC endpoint" />
              )}
            />
          </Form>
        )}
      />
    )
  }
}

const mapStateToProps = (state: AppState) => {
  return {
    createConnection: state.createConnection,
    visible: state.createConnection.status === Status.Started,
    submitted: state.createConnection.status === Status.Completed,
    loading: state.createConnection.status === Status.InProgress
  }
}

const mapDispatchToProps = (dispatch: Dispatch) => {
  return bindActionCreators(
    {
      createOrUpdateConnection,
      createConnectionCancelled
    },
    dispatch
  )
}

export default connect<StateProps, DispatchProps, {}, AppState>(
  mapStateToProps,
  mapDispatchToProps
)(ConnectionModal)
