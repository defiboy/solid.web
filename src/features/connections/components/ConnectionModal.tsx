import React from 'react'

import { bindActionCreators, Dispatch, ActionCreator, Action } from 'redux'
import { connect } from 'react-redux'
import { Form } from 'antd'
import { FormikErrors, Field, FieldProps } from 'formik'
import axios from 'axios'
import { Connection } from '@solid-explorer/types'

import { ApplicationState } from '../../rootReducer';
import { InputFormItem } from '../../../components'
import { Status } from "../../common/types" // TODO: this shouldn't be the case with Sagas

import { createOrUpdateConnection, closeConnectionModal } from '../actions'

import { ConnectionModalComponent } from "./ConnectionModalComponent";
import { RadioField } from './RadioField'
import { ConnectionType, PublicChainId } from '@solid-explorer/types/lib/connections/ConnectionType'
import { SelectField } from './SelectField'

const FORM_ID = 'CONNECTION_FORM'
const FORM_TITLE = "Add Connection"

const defaultConnection: Connection = {
  name: '',
  url: '',
  type: ConnectionType.Private,
  publicChainId: PublicChainId.Goerli
}

const defaultRadioFormOptions = [
  {
    key: ConnectionType.Private,
    value: ConnectionType.Private
  },
  {
    key: ConnectionType.Public,
    value: ConnectionType.Public
  }
]

const defaultPublicChainFormOptions = [
  {
    key: PublicChainId.Goerli,
    value: PublicChainId.Goerli
  },
  {
    key: PublicChainId.MainNet,
    value: PublicChainId.MainNet
  },
  {
    key: PublicChainId.XDai,
    value: PublicChainId.XDai
  }
]

interface OwnProps {
  item: Connection
}

interface StateProps {
  visible: boolean
  loading: boolean
  submitted: boolean
}

interface DispatchProps {
  createOrUpdateConnection: (item: Connection) => void
  closeConnectionModal: ActionCreator<Action>
}

type AllProps = OwnProps & DispatchProps & StateProps

export class ConnectionModal extends React.Component<AllProps> {

  static defaultProps = {
    item: defaultConnection
  }

  saveConnection = async (item: Connection, actions: any) => {
    console.log("ITEM", item)
    console.log("ACTIONS", actions)
    const isValid = await isValidConnection(item.url)
    if (!isValid) {
      actions.setSubmitting(true)
      actions.setFieldError('url', 'Invalid JSON RPC URL')
    } else {
      // TODO test.. to avoid store data that can confuse...
      const itemCleaned: Connection = item.type === ConnectionType.Private ? { ...item, publicChainId: undefined } : { ...item, url: '' }
      this.props.createOrUpdateConnection(itemCleaned)
    }
  }

  render() {
    return (
      <ConnectionModalComponent
        formId={FORM_ID}
        title={FORM_TITLE}
        visible={this.props.visible}
        loading={this.props.loading}
        initialValues={this.props.item}
        onSubmit={this.saveConnection}
        onCancel={this.props.closeConnectionModal}
        validator={validator}
        FormComponent={({ onSubmit }) => (
          <Form id={FORM_ID} onSubmit={onSubmit}>
            <Field
              name="name"
              render={(innerProps: FieldProps) => <InputFormItem {...innerProps} label="Name" placeHolder="name" />}
            />
            <Field
              name="type"
              render={(innerProps: FieldProps) => (
                <RadioField isButton={true} options={defaultRadioFormOptions} defaultValue={ConnectionType.Private} label="Type" {...innerProps} />
              )}
            />
            <Field
              name="publicChainId"
              as="select"
              render={(innerProps: FieldProps) => (
                <SelectField {...innerProps} options={defaultPublicChainFormOptions} label="Public network" />
              )}
            />
            <Field
              name="url"
              render={(innerProps: FieldProps) => {
                const isPrivate = innerProps.form.values.type === ConnectionType.Private
                return isPrivate ? <InputFormItem {...innerProps} label="Blockchain URL" placeHolder="JSON RPC endpoint" /> : <div></div>
              }}
            />
          </Form>
        )}
      />
    )
  }
}

// TODO TO MOVE
const isValidConnection = async (url: string) => {
  try {
    const result = await axios.post(url, '{"jsonrpc":"2.0","method":"web3_clientVersion","params":[],"id":67}')
    return result.data;
  } catch (error) {
    return undefined;
  }
}

const validator = (items: Connection): FormikErrors<Connection> => {
  const errors: FormikErrors<Connection> = {}
  console.log("ITEMS", items)
  if (!items.name) {
    errors.name = 'Required'
  }
  if (!items.url) {
    errors.url = 'Required'
  }
  if (items.type === ConnectionType.Public && !items.publicChainId) {
    errors.publicChainId = 'Required'
  }
  return errors
}

const mapStateToProps = ({ connectionState }: ApplicationState) => {
  return {
    visible: connectionState.connectionModalOpen,
    submitted: connectionState.createConnectionStatus === Status.Completed,
    loading: connectionState.createConnectionStatus === Status.InProgress
  }
}

const mapDispatchToProps = (dispatch: Dispatch) => {
  return bindActionCreators(
    {
      createOrUpdateConnection,
      closeConnectionModal
    },
    dispatch
  )
}

export default connect<StateProps, DispatchProps, {}, ApplicationState>(
  mapStateToProps,
  mapDispatchToProps
)(ConnectionModal)
