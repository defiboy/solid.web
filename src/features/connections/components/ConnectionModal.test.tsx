import '@testing-library/jest-dom/extend-expect'

import React from 'react';
import { FormikErrors, Field, FieldProps } from 'formik';
import { render } from '@testing-library/react'
import { Form } from 'antd'
import { Provider } from 'react-redux'
import { createStore } from 'redux'

import { ConnectionModal, cleanDataBeforeSubmit } from './ConnectionModal'
import rootReducer, { ApplicationState } from '../../rootReducer'
import { buildFakeConnection } from '@solid-explorer/types';

describe('GenericModal', () => {

    const MODAL_TITLE = "modal title"
    const FORM_ID = "modalid"

    const onSubmitMockHandler = jest.fn()
    const onCancelMockHandler = jest.fn()
    const onCreateOrUpdateConnectionMockHandler = jest.fn()
    const onCloseConnectionModalMockHandler = jest.fn()

    const renderConnectionModal = () => {
        const store = createStore(rootReducer)

        return render(
            <Provider store={store}>
                <ConnectionModal visible={true} loading={false}
                    submitted={false}
                    createOrUpdateConnection={jest.fn()}
                    closeConnectionModal={jest.fn()}
                />
            </Provider>
        )
    }

    it.skip('renders a modal with basic elements', () => {
        const { getByLabelText, getByTestId } = renderConnectionModal()

        expect(getByLabelText(`${MODAL_TITLE}`)).toBeInTheDocument()
        expect(getByTestId('input-name')).toBeInTheDocument()
        expect(getByTestId('input-url')).toBeInTheDocument()
        expect(getByTestId(`submit-button-${FORM_ID}`)).toBeInTheDocument()
        expect(getByTestId(`cancel-button-${FORM_ID}`)).toBeInTheDocument()
    })

    it.skip('renders a modal', () => {
        const { baseElement } = renderConnectionModal()

        expect(baseElement).toMatchSnapshot()
    })

    // TODO
    // describe("cleanDataBeforeSubmit", () => {
    //     it('should return a valid private connection', () => {
    //         const connection = buildFakeConnection({
    //             type: ConnectionType.Private,

    //         })
    //         const expectedConnection = {
    //             ...connection
    //         }
    //         const result = cleanDataBeforeSubmit(connection)
    //     })

    //     it('should return a valid public connection', () => {

    //     })
    // })

})
