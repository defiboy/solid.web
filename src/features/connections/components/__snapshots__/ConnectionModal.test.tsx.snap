// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericModal renders a modal 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div>
    <div>
      <div
        class="ant-modal-mask fade-appear"
      />
      <div
        aria-labelledby="rcDialogTitle0"
        class="ant-modal-wrap "
        role="dialog"
        tabindex="-1"
      >
        <div
          class="ant-modal zoom-appear"
          role="document"
          style="width: 500px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <button
              aria-label="Close"
              class="ant-modal-close"
              type="button"
            >
              <span
                class="ant-modal-close-x"
              >
                <i
                  aria-label="icon: close"
                  class="anticon anticon-close ant-modal-close-icon"
                >
                  <svg
                    aria-hidden="true"
                    class=""
                    data-icon="close"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                    />
                  </svg>
                </i>
              </span>
            </button>
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="rcDialogTitle0"
              >
                Add Connection
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <div>
                <form
                  class="ant-form ant-form-horizontal"
                  id="CONNECTION_FORM"
                >
                  <div
                    class="ant-row ant-form-item"
                  >
                    <div
                      class="ant-col ant-form-item-label"
                    >
                      <label
                        class=""
                        for="name"
                        title="Name"
                      >
                        Name
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control-wrapper"
                    >
                      <div
                        class="ant-form-item-control"
                      >
                        <span
                          class="ant-form-item-children"
                        >
                          <input
                            class="ant-input ant-input-lg"
                            data-testid="input-name"
                            id="name"
                            name="name"
                            placeholder="name"
                            type="text"
                            value=""
                          />
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-row ant-form-item"
                  >
                    <div
                      class="ant-col ant-form-item-label"
                    >
                      <label
                        class=""
                        title="Type"
                      >
                        Type
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control-wrapper"
                    >
                      <div
                        class="ant-form-item-control"
                      >
                        <span
                          class="ant-form-item-children"
                        >
                          <div
                            class="ant-radio-group ant-radio-group-outline"
                          >
                            <label
                              class="ant-radio-button-wrapper ant-radio-button-wrapper-checked"
                            >
                              <span
                                class="ant-radio-button ant-radio-button-checked"
                              >
                                <input
                                  checked=""
                                  class="ant-radio-button-input"
                                  name="type"
                                  type="radio"
                                  value="Private"
                                />
                                <span
                                  class="ant-radio-button-inner"
                                />
                              </span>
                              <span>
                                Private
                              </span>
                            </label>
                            <label
                              class="ant-radio-button-wrapper"
                            >
                              <span
                                class="ant-radio-button"
                              >
                                <input
                                  class="ant-radio-button-input"
                                  name="type"
                                  type="radio"
                                  value="Public"
                                />
                                <span
                                  class="ant-radio-button-inner"
                                />
                              </span>
                              <span>
                                Public
                              </span>
                            </label>
                          </div>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div />
                  <div
                    class="ant-row ant-form-item"
                  >
                    <div
                      class="ant-col ant-form-item-label"
                    >
                      <label
                        class=""
                        for="url"
                        title="Blockchain URL"
                      >
                        Blockchain URL
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control-wrapper"
                    >
                      <div
                        class="ant-form-item-control"
                      >
                        <span
                          class="ant-form-item-children"
                        >
                          <input
                            class="ant-input ant-input-lg"
                            data-testid="input-url"
                            id="url"
                            name="url"
                            placeholder="JSON RPC endpoint"
                            type="text"
                            value=""
                          />
                        </span>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
            <div
              class="ant-modal-footer"
            >
              <button
                class="ant-btn"
                data-testid="cancel-button-CONNECTION_FORM"
                type="button"
              >
                <span>
                  Cancel
                </span>
              </button>
              <button
                class="ant-btn ant-btn-primary"
                data-testid="submit-button-CONNECTION_FORM"
                form="CONNECTION_FORM"
                type="submit"
              >
                <span>
                  Save
                </span>
              </button>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
